File: sdk/python/src/soulfire/actions.py
from .bot_live_pb2 import (
    BOT_ACTION_STATUS_COMPLETED,
    BotActionResult,
)


class SoulFireActionError(RuntimeError):
    def __init__(self, result: BotActionResult) -> None:
        self.result = result
        super().__init__(result.error or f"Bot action {result.action_id} did not complete")


def require_action(result: BotActionResult) -> BotActionResult:
    if result.status != BOT_ACTION_STATUS_COMPLETED:
        raise SoulFireActionError(result)
    return result


def action_headers(
    headers: dict[str, str] | None,
    token: str | None,
) -> dict[str, str] | None:
    if token is None:
        return headers
    result = dict(headers or {})
    result["X-SoulFire-Control-Token"] = token
    return result

File: sdk/typescript/src/actions.ts
import {
  BotActionStatus,
  type BotActionResult,
} from "./generated/soulfire/bot_live_pb.js";

export class SoulFireActionError extends Error {
  public constructor(public readonly result: BotActionResult) {
    super(result.error ?? `Bot action ${result.actionId} did not complete`);
    this.name = "SoulFireActionError";
  }
}

export function requireCompletedAction(
  result: BotActionResult | undefined,
): BotActionResult {
  if (result === undefined) {
    throw new Error("SoulFire did not return a bot action result");
  }
  if (result.status !== BotActionStatus.COMPLETED) {
    throw new SoulFireActionError(result);
  }
  return result;
}
